core/embed/prodtest/main.c

```diff
@@ -513,9 +603,446 @@ static void test_otp_write_device_variant(const char *args) {
+  cpuid[0] = LL_GetUID_Word0();
+  cpuid[1] = LL_GetUID_Word1();
+  cpuid[2] = LL_GetUID_Word2();
+
+  vcp_print("OK ");
+  vcp_println_hex((uint8_t *)cpuid, sizeof(cpuid));
+}
+
+#ifdef USE_OPTIGA
+static const uint16_t OID_CERT_INF = 0xE0E0;
+static const uint16_t OID_CERT_DEV = 0xE0E1;
+static const uint16_t OID_CERT_FIDO = 0xE0E2;
+static const uint16_t OID_KEY_DEV = 0xE0F0;
+static const uint16_t OID_KEY_FIDO = 0xE0F2;
+static const uint16_t OID_KEY_PAIRING = 0xE140;
+static const uint16_t OID_OPTIGA_UID = 0xE0C2;
+static const uint16_t OID_TRUST_ANCHOR = 0xE0E8;
+
+// Data object access conditions.
+static const optiga_metadata_item ACCESS_PAIRED = {
+    (const uint8_t *)"\x20\xE1\x40", 3};
+static const optiga_metadata_item KEY_USE_SIGN = {(const uint8_t *)"\x10", 1};
+static const optiga_metadata_item TYPE_PTFBIND = {(const uint8_t *)"\x22", 1};
+
+static bool set_metadata(uint16_t oid, const optiga_metadata *metadata) {
+  uint8_t serialized[258] = {0};
+  size_t size = 0;
+  optiga_result ret = optiga_serialize_metadata(metadata, serialized,
+                                                sizeof(serialized), &size);
+  if (OPTIGA_SUCCESS != ret) {
+    vcp_println("ERROR optiga_serialize_metadata error %d for OID 0x%04x.", ret,
+                oid);
+    return false;
+  }
+
+  optiga_set_data_object(oid, true, serialized, size);
+
+  ret =
+      optiga_get_data_object(oid, true, serialized, sizeof(serialized), &size);
+  if (OPTIGA_SUCCESS != ret) {
+    vcp_println("ERROR optiga_get_metadata error %d for OID 0x%04x.", ret, oid);
+    return false;
+  }
+
+  optiga_metadata metadata_stored = {0};
+  ret = optiga_parse_metadata(serialized, size, &metadata_stored);
+  if (OPTIGA_SUCCESS != ret) {
+    vcp_println("ERROR optiga_parse_metadata error %d.", ret);
+    return false;
+  }
+
+  if (!optiga_compare_metadata(metadata, &metadata_stored)) {
+    vcp_println("ERROR optiga_compare_metadata failed.");
+    return false;
+  }
+
+  return true;
+}
+
+static bool pair_optiga(void) {
+  // The pairing key may already be written and locked. The success of the
+  // pairing procedure is determined by optiga_sec_chan_handshake(). Therefore
+  // it is OK for some of the intermediate operations to fail.
+
+  // Enable writing the pairing secret to OPTIGA.
+  optiga_metadata metadata = {0};
+  metadata.change = OPTIGA_ACCESS_ALWAYS;
+  metadata.execute = OPTIGA_ACCESS_ALWAYS;
+  metadata.data_type = TYPE_PTFBIND;
+  set_metadata(OID_KEY_PAIRING, &metadata);  // Ignore result.
+
+  // Generate pairing secret.
+  uint8_t secret[SECRET_OPTIGA_KEY_LEN] = {0};
+  optiga_result ret = optiga_get_random(secret, sizeof(secret));
+  if (OPTIGA_SUCCESS != ret) {
+    vcp_println("ERROR optiga_get_random error %d,", ret);
+    return false;
+  }
+
+  // Store pairing secret.
+  ret = optiga_set_data_object(OID_KEY_PAIRING, false, secret, sizeof(secret));
+  /*
+   * TODO: Uncomment. Right now this code will render the device unusable with
```

Comment:
please uncomment this now, with the modifications in bootloader this should no longer be needed

Reply:
f92db56fc8319aa573536e39396597e0e99ebd46